Rework sam file tutorial (#2610)
seqan/seqan3: doc/tutorial/sam_file/index.md

in which case all template parameters are automatically deduced (by the file name extension).
Or you can pass a stream (e.g. std::cin or std::stringstream), but then you need to know your format beforehand:

\snippet doc/tutorial/sam_file/sam_file_filename_construction.cpp main

> ## Using the seqan3::sam_record

Author: ```suggestion
## Accessing individual record members
```

You can access the record member like this

\snippet doc/tutorial/sam_file/sam_file_sam_record.cpp main

See seqan3::sam_record for all data accessor.

\cond DEV

> \todo What to do with this section? We want to advertise for seqan3.1 to use records.

Author: This was more a comment for enrico and to remind myself :)

> \todo What to do with this section? We want to advertise for seqan3.1 to use records.

Author: We decided to remove this section

## Structural bindings

In many cases you are not interested in all of the information in a file to be stored.
For this purpose, we provide the possibility to select specific seqan3::field's for a file.
The file will read only those fields and fill the record accordingly.

You can select fields by providing a seqan3::fields object as an extra parameter to the constructor:

\snippet doc/tutorial/sam_file/sam_file_read_custom_fields.cpp main

\attention The order in which you specify the selected fields determines the order of elements in the

>            seqan3::sam_record.

Author: like here

you can provide these information to the seqan3::sam_file_input which automatically fills the alignment object.
You can pass reference ids and reference sequences as additional constructor parameters:

\snippet doc/tutorial/sam_file/sam_file_alignments_with_ref.cpp main

> 

Reviewer: ```suggestion
The code will print the following:
\include doc/tutorial/sam_file/sam_file_sam_record.out

```

\assignment{Assignment 2: Combining sequence and alignment files}

> Read in the following reference sequence FASTA file (see the sequence file tutorial if you need a reminder):

Reviewer: ```suggestion
Read the following reference sequence FASTA file (see the sequence file tutorial if you need a reminder):
```

\snippet doc/tutorial/sam_file/sam_file_solution2.cpp ref_file

> Then read the following SAM file while providing the reference sequence information.
> 

Reviewer: ```suggestion
Then read the following SAM file while providing the reference sequence information.

\snippet doc/tutorial/sam_file/sam_file_solution2.cpp sam_file

```

Only use
* seqan3::sam_record::id,
* seqan3::sam_record::reference_id,
* seqan3::sam_record::mapping_quality, and
* seqan3::sam_record::alignment.

> 
> \snippet doc/tutorial/sam_file/sam_file_solution2.cpp sam_file

Reviewer: ```suggestion
```

>   * For the resulting alignments, print which read was mapped against with reference id and

Author: @eseiler can you help?

>   * For the resulting alignments, print which read was mapped against with reference id and

Reviewer: The suggestion is correct

>     the number of seqan3::gap's in each sequence (aligned reference and read sequence).

Reviewer: ```suggestion
    the number of `seqan3::gap`s in each sequence (aligned reference and read sequence).
```

Would this look better? ' is not correct :)

\cond DEV
\todo This isn't used, right? Maybe a left over? seqan3::sam_record::reference_id?! Maybe we should make the example
      better to understand that comment.

\note reference ids (field::ref_id) are given as an index of type `std::optional<int32_t>`

>       that denote the position of the reference id in the `ref_ids` vector passed to the alignment file.

Author: Yeah, I hope that @eseiler can remember xD

But I think `record.reference_id()` is of type `std::optional<int32_t>`.

The problem is that the debug stream does not need `optional.get()` to access the data (as it does that automatically).

So I guess 

```
std::optional reference_id = record.reference_id();

seqan3::debug_stream << "refence_id: " << reference_id.get();
```

might be a better example

>       that denote the position of the reference id in the `ref_ids` vector passed to the alignment file.

Reviewer: Ah, yeah, just do `reference_id.get()` and (maybe) as a note that reference_id is an optional, but no explanation as to what this represents ( the `that denote the position of the reference id in the `ref_ids` vector passed to the alignment file.` part)
